### PR TITLE
Clean export tempfiles on failure

### PR DESCRIPTION
**Pull request by esoergel**
Best reviewed :tropical_fish: :heavy_multiplication_x: :tropical_fish: 
@mkangia
@emord @snopoke potentially interested parties

### CODE REVIEW

File: corehq/util/files.py [hunk 1]
```diff
@@ -40,3 +42,25 @@ def safe_filename_header(filename, extension=None):
...
+            # path exists throughout this block
+            with open(path, 'w') as f:
+                f.write("Adding stuff to the file")
+            with open(path) as f:
+                do_stuff(f.read())
+        # path no longer exists
+    """
+
+    def __enter__(self):
+        _, self.path = tempfile.mkstemp()
```
Comment (millerdev) [inline corehq/util/files.py]: I think you want to call `.close()` on thing you're throwing away here (`_`).

File: corehq/apps/export/tests/test_export_form_subcases.py [hunk 2]
```diff
@@ -15,7 +15,7 @@
 from corehq.apps.app_manager.signals import app_post_save
 from corehq.apps.export.dbaccessors import delete_all_export_data_schemas
 from corehq.apps.export.models import FormExportDataSchema, FormExportInstance
-from corehq.apps.export.export import get_export_file
+from corehq.apps.export.tests.util import get_export_json
```
Comment (esoergel) [inline corehq/apps/export/tests/test_export_form_subcases.py]: This commit is best viewed with whitespace ignored:
https://github.com/dimagi/commcare-hq/pull/20659/commits/6d55dfd8576d5f7161534ac4c5b1bc86a40af282?w=1

File: corehq/apps/export/export.py [hunk 3]
```diff
@@ -48,19 +47,18 @@ def __enter__(self):
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.file.close()
-        os.remove(self.path)
```
Comment (esoergel) [inline corehq/apps/export/export.py]: This is the old way these were cleaned up - via `with` statement used when the generated file was read.  Removing this because `TransientTempfile` handles the cleanup now.

**Review by emord — COMMENTED** [review]
How hard would it be to change the code to use `tempfile.TemporaryFile` ? From what I can tell the difference is that your helper provides a path and the official temporary file provides a file like object.

(also nit that `TransientTempfile` feels a little redundant)

standard lib: https://docs.python.org/2/library/tempfile.html#tempfile.TemporaryFile

**Comment by esoergel** [comment]
> How hard would it be to change the code to use tempfile.TemporaryFile ? From what I can tell the difference is that your helper provides a path and the official temporary file provides a file like object.

I commented about that a bit in my commit message: https://github.com/dimagi/commcare-hq/pull/20659/commits/f34f48e98bd9c054865b920153a3ef2638526444
`NamedTemporaryFile` provides a path and otherwise has the same API as `TemporaryFile`.  The problem is both of them automatically delete the file when it's closed (although you can make `NamedTemporaryFile` persist if you want).  To use one of those, the code would probably use a `with` statement at the same places as done here, and pass around a file object rather than a path.  The file object would be open for reading and writing.  When the writes are finished, it'd need to `seek` back to the start of the file before the reads begin.

In short, that would probably work too, I just thought this presented a nicer interface, since we can read and write in isolation and close the file in between, which felt more in line with how we typically access files.

> also nit that `TransientTempfile` feels a little redundant

Do you mean in name?  I agree, and am open to suggestions!

**Comment by emord** [comment]
>  commented about that a bit in my commit message: f34f48e
ack. I think I read that and then immediately forgot about it. That makes sense.

Maybe `ReadableTemporaryFile`? Not the greatest, but it hints at the benefit of using this class over the standard library.

File: corehq/util/files.py (hunk 1 above)
Comment (esoergel) [reply corehq/util/files.py]: Oh yikes, you're right.  We do this all over the place.  Updated.

File: corehq/util/files.py [hunk 4]
```diff
@@ -59,7 +59,8 @@ class TransientTempfile(object):
     """
 
     def __enter__(self):
-        _, self.path = tempfile.mkstemp()
+        fd, self.path = tempfile.mkstemp()
+        os.close(fd)
```
Comment (millerdev) [inline corehq/util/files.py]: There's a smell here because this is closing, and then inside the context manager immediately re-opening the same file. From a resource management perspective this might be more efficient:

```py
@attr.s
class TransientTempfile(object):
    """
    Manage a temporary file that can be opened and closed as needed, but will
    be automatically cleaned up on exit.

        with TransientTempfile() as fileobj, path:
            # path exists throughout this block
            with fileobj:
                fileobj.write("Adding stuff to the file")
            with open(path) as f:
                do_stuff(f.read())
        # path no longer exists
    """

    mode = attr.ib(default="wt")
    encoding = attr.ib(default="utf-8")

    def __enter__(self):
        fd, self.path = tempfile.mkstemp()
        self.encoding = None if "b" in self.mode else self.encoding
        fileobj = io.open(fd, self.mode, encoding=self.encoding)
        return fileobj, self.path

    def __exit__(self, exc_type, exc_val, exc_tb):
        os.remove(self.path)
```

Maybe this makes the implementation too complex? Your call if you want to use it or not.

File: corehq/apps/export/export.py [hunk 5]
```diff
@@ -336,19 +309,19 @@ def get_export_file(export_instances, filters, progress_tracker=None):
...
 
     def iter_export_docs():
-        # Stream doc ids from disk and fetch documents from ES in chunks
-        with open(temp_path) as f:
-            doc_ids = (doc_id.strip() for doc_id in f)
-            for doc in iter_es_docs(query.index, doc_ids):
-                yield doc
-        os.remove(temp_path)
+        with TransientTempfile() as temp_path:
+            with open(temp_path, 'w') as f:
```
Comment (millerdev) [inline corehq/apps/export/export.py]: I think this is using the default `open` when it should be using `io.open` and passing in `encoding="utf-8"`. Same below on line 321.

File: corehq/apps/export/multiprocess.py [hunk 6]
```diff
@@ -185,12 +185,15 @@ def _doc_iter():
     return ScanResult(doc_count, _doc_iter())
 
 
-def get_export_file(export_instance, docs, progress_tracker=None):
+def _get_export_file_path(export_instance, docs, progress_tracker=None):
     export_instances = [export_instance]
-    writer = get_export_writer(export_instances, allow_pagination=False)
+    # Multiprocess exports sometimes intentionally keep the tempfile,
+    # so TransientTempfile isn't appropriate here
+    _, temp_path = tempfile.mkstemp()
```
Comment (millerdev) [inline corehq/apps/export/multiprocess.py]: should close file descriptor (`_`)

**Review by millerdev — COMMENTED** [review]
Didn't do a complete review, but made a few comments related to opening files with text encoding. Not sure if some of this will work on Python 3 without being more explicit about encodings.

File: corehq/util/files.py (hunk 4 above)
Comment (esoergel) [reply corehq/util/files.py]: I thought about making this util manage opening and closing the tempfile as well, but I like the idea of keeping it responsible solely for allocating the tempfile and not making a new api for file access.  I think I'll keep it as-is for now.

File: corehq/apps/export/multiprocess.py (hunk 6 above)
Comment (esoergel) [reply corehq/apps/export/multiprocess.py]: updated

File: corehq/apps/export/export.py (hunk 5 above)
Comment (esoergel) [reply corehq/apps/export/export.py]: updated

File: corehq/apps/export/export.py [hunk 7]
```diff
@@ -336,19 +310,19 @@ def get_export_file(export_instances, filters, progress_tracker=None):
...
-        # Stream doc ids from disk and fetch documents from ES in chunks
-        with open(temp_path) as f:
-            doc_ids = (doc_id.strip() for doc_id in f)
-            for doc in iter_es_docs(query.index, doc_ids):
-                yield doc
-        os.remove(temp_path)
+        with TransientTempfile() as temp_path:
+            with open(temp_path, 'w', encoding='utf-8') as f:
+                for doc_id in scroll_result:
+                    f.write(doc_id + '\n')
```
Comment (mkangia) [inline corehq/apps/export/export.py]: would it be better to write at once?
`f.write('\n'.join(doc_id for doc_id in scroll_result))`

This would probably depend on the number of ids here. if its just a few or too large this might be not-much-difference/too-much-memory but does avoid that many `write` method calls.

File: corehq/apps/export/tasks.py [hunk 8]
```diff
@@ -24,27 +24,29 @@
...
-            mimetype=file_format.mimetype,
-            content_disposition=safe_filename_header(filename, file_format.extension),
-            download_id=download_id,
-        )
+            expose_blob_download(
+                download_id,
+                mimetype=file_format.mimetype,
+                content_disposition=safe_filename_header(filename, file_format.extension),
+                download_id=download_id,
+            )
```
Comment (mkangia) [inline corehq/apps/export/tasks.py]: Can use `ExposeBlobDownload` here to do some of this? Probably out of scope of this PR so feel free to skip it if needed.

**Review by mkangia — APPROVED** [review]
Couple of minor things.

File: corehq/apps/export/export.py (hunk 7 above)
Comment (esoergel) [reply corehq/apps/export/export.py]: It's potentially a very big list, and I think that approach would require pulling it all into memory and committing in one go.  Perhaps a chunked approach would be better?  I'll call that out of scope for this PR, since I wasn't intending to change what happens with the tempfiles, just that they are guaranteed to get cleaned up.

File: corehq/apps/export/export.py (hunk 5 above)
Comment (esoergel) [reply corehq/apps/export/export.py]: Just reverted that change because it caused build failures.  I'm just gonna leave these as they were beforehand and deal with making this code python 3 compatible in another PR.
